torch/_functorch/_aot_autograd/runtime_wrappers.py

```diff
@@ -896,7 +897,7 @@ def _same_dtype_views(view1, view2):
     base_args = []
     other_args = []
     for i, inpt in enumerate(fwd_inputs):
-        if isinstance(inpt, Tensor):
+        if isinstance(inpt, Tensor) and not is_sparse_any(inpt):
```

Comment:
Same here, it should be a TODO item that involves dealing with sparse tensors `values` storage.

What are the test failures when disabling `_aot_autograd` changes of this PR?

Reply:
I'm curious how you ended up here - did the basic tests added in this PR end up going down this code path?

This code path is extremely rare / not very well hardened, because it covers the specific situation where you have a graph with two tensor inputs that alias one another, and at least one of those inputs is mutated. So it seems safer to blanket raise an error whenever any of our inputs are sparse tensors here instead of trying to handle it.

Reply:
Maybe a more general question/comment: I'm not sure what will happen today when you plumb sparse FakeTensors through functionalization. Are there any tests to see if any thing reasonable happens?

A safe first step might be to blanket ban all mutations on sparse tensors in functionalization, as a way to make sure we aren't allowing silent correctness problem. If that sounds reasonable, I can point to the right places to add an error like that.

Reply:
I must admit that I am still not fully up to speed to answer all questions. When I run the tests "as is" (i.e. just for COO, since right now we skip all non-COO anyway), I already hit this point with the following stack trace. Note that all remaining issues (also for the non-COO sparsity that is currently skipped) seem all related to the missing functionalization. I of course plan to work on that next (but I was just hoping to get this very basic test in as a landing place for next improvements). If you prefer we fix everything first (including the non-COO), I understand that of course fully and would love to get some pointers on sending out smaller PRs to fix issues one by one first.

```
  File "ajcbik/PYTORCH/pytorch/test/export/test_sparse.py", line 131, in test_sumnet
    prog = torch.export.export(net, (sparse_input,))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/export/__init__.py", line 174, in export
    return _export(
           ^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/export/_trace.py", line 834, in wrapper
    raise e
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/export/_trace.py", line 817, in wrapper
    ep = fn(*args, **kwargs)
         ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/export/exported_program.py", line 85, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/export/_trace.py", line 1168, in _export
    ep_non_strict = _export_non_strict(
                    ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/export/_trace.py", line 521, in _export_non_strict
    gm, graph_signature = transform(aot_export_module)(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/_functorch/aot_autograd.py", line 1135, in aot_export_module
    fx_g, metadata, in_spec, out_spec = _aot_export_function(
                                        ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/_functorch/aot_autograd.py", line 1354, in _aot_export_function
    fx_g, meta = create_aot_dispatcher_function(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/_dynamo/utils.py", line 267, in time_wrapper
    r = func(*args, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/_functorch/aot_autograd.py", line 685, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(
                  ^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 471, in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 666, in aot_wrapper_synthetic_base
    flat_args_with_synthetic_bases, synthetic_base_info = merge_view_inputs(
                                                          ^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 902, in merge_view_inputs
    storage_ref = StorageWeakRef(inpt.untyped_storage())
```

Reply:
I suggest the following change:
```
diff --git a/torch/_functorch/_aot_autograd/runtime_wrappers.py b/torch/_functorch/_aot_autograd/runtime_wrappers.py
index dad7dd9290f..5c9409096e1 100644
--- a/torch/_functorch/_aot_autograd/runtime_wrappers.py
+++ b/torch/_functorch/_aot_autograd/runtime_wrappers.py
@@ -894,12 +894,15 @@ def merge_view_inputs(
         return True
 
     assert len(fwd_inputs) == len(mutated_input_info)
+    if not [info for info in mutated_input_info if info.mutates_data]:
+        # return early when there are no mutations
+        return fwd_inputs, None
+
     storage_ref_to_idx: Dict[StorageWeakRef, List[int]] = collections.defaultdict(list)
     base_args = []
     other_args = []
     for i, inpt in enumerate(fwd_inputs):
-        # TODO: remove is_sparse_any clause by supporting storage?
-        if isinstance(inpt, Tensor) and not is_sparse_any(inpt):
+        if isinstance(inpt, Tensor):
             storage_ref = StorageWeakRef(inpt.untyped_storage())
             storage_ref_to_idx[storage_ref].append(i)
         else:
```
that skips input processing that turns out to be unnecessary when no mutations exists. It will also postpone handling possibly mutated sparse tensors until we'll have a real use case for it that would fail here so that we could cook up a working solution in the future.

@bdhirsh what to you think?